src/data/sum/order.lean

```diff
@@ -0,0 +1,511 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import order.hom.basic
+import order.lexicographic
+
+/-!
+# Orders on a sum type
+
+This file defines the disjoint sum and the linear (aka lexicographic) sum of two orders and provides
+relation instances for `sum.lift_rel` and `sum.lex`.
+
+We declare the disjoint sum of orders as the default set of instances. The linear order goes on a
+type synonym.
+
+## Main declarations
+
+* `sum.has_le`, `sum.has_lt`: Disjoint sum of orders.
+* `sum.lex.has_le`, `sum.lex.has_lt`: Lexicographic/linear sum of orders.
+
+## Notation
+
+* `α ⊕ₗ β`:  The linear sum of `α` and `β`.
+-/
+
+namespace sum
+variables {α β γ δ : Type*}
+
+/-! ### Unbundled relation classes -/
+
+section lift_rel
+variables (r : α → α → Prop) (s : β → β → Prop)
+
+instance [is_refl α r] [is_refl β s] : is_refl (α ⊕ β) (lift_rel r s) :=
+⟨by { rintro (a | a), exacts [lift_rel.inl (refl _), lift_rel.inr (refl _)] }⟩
+
+instance [is_irrefl α r] [is_irrefl β s] : is_irrefl (α ⊕ β) (lift_rel r s) :=
+⟨by { rintro _ (⟨a, _, h⟩ | ⟨a, _, h⟩); exact irrefl _ h }⟩
+
+instance [is_trans α r] [is_trans β s] : is_trans (α ⊕ β) (lift_rel r s) :=
+⟨by { rintro _ _ _ (⟨a, b, hab⟩ | ⟨a, b, hab⟩) (⟨_, c, hbc⟩ | ⟨_, c, hbc⟩),
+  exacts [lift_rel.inl (trans hab hbc), lift_rel.inr (trans hab hbc)] }⟩
+
+instance [is_antisymm α r] [is_antisymm β s] : is_antisymm (α ⊕ β) (lift_rel r s) :=
+⟨by { rintro _ _ (⟨a, b, hab⟩ | ⟨a, b, hab⟩) (⟨_, _, hba⟩ | ⟨_, _, hba⟩); rw antisymm hab hba }⟩
```

Comment:
IMO these would be clearer if written with `match`, even though that makes them longer.

It's worth having separate `refl` and `is_refl` definitions too, so that you can tag the former with `@[refl]`. The same is true of `trans`, but probably not for the others.